server/comms/rpc/rpctypes.go

```diff
@@ -0,0 +1,358 @@
+	Payload json.RawMessage `json:"payload",omitempty`
+}
+
+// DecodeMessage decodes a *Message from JSON-formatted bytes.
+func DecodeMessage(b []byte) (*Message, error) {
+	msg := new(Message)
+	err := json.Unmarshal(b, &msg)
+	if err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+// NewRequest is the constructor for a Request-type *Message.
+func NewRequest(id uint64, route string, payload interface{}) (*Message, error) {
+	encoded, err := json.Marshal(payload)
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type:    Request,
+		Payload: json.RawMessage(encoded),
+		Route:   route,
+		ID:      id,
+	}, nil
+}
+
+// NewResponse encodes the result and creates a Response-type *Message.
+func NewResponse(id uint64, result interface{}, rpcErr *Error) (*Message, error) {
+	encResult, err := json.Marshal(result)
+	if err != nil {
+		return nil, err
+	}
+	resp := &ResponsePayload{
+		Result: encResult,
+		Error:  rpcErr,
+	}
+	encResp, err := json.Marshal(resp)
+	if err != nil {
+		return nil, err
+	}
+	return &Message{
+		Type:    Response,
+		Payload: json.RawMessage(encResp),
+		ID:      id,
+	}, nil
+}
+
+// Response attempts to decode the payload to a *ResponsePayload. Response will
+// return an error if the Type is not Response.
+func (msg *Message) Response() (*ResponsePayload, error) {
+	if msg.Type != Response {
+		return nil, fmt.Errorf("invalid type %d for ResponsePayload", msg.Type)
+	}
+	resp := new(ResponsePayload)
+	err := json.Unmarshal(msg.Payload, &resp)
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+// MatchNotification is the params for a DEX-originating MatchRoute request.
+type Match struct {
+	signable
+	OrderID  Bytes `json:"orderid"`
+	MatchID  Bytes `json:"matchid"`
+	Quantity uint64 `json:"quantity"`
+	Rate     uint64 `json:"rate"`
+	Address  string `json:"address"`
+	Time     uint64 `json:"timestamp"`
+}
+
+var _ Signable = (*Match)(nil)
+
+// Serialize serializes the Match data.
+func (m *Match) Serialize() ([]byte, error) {
+	// Match serialization is orderid (32) + matchid (32) + quantity (8) + rate (8)
+	// + address (variable, guess 35). Sum = 115
+	s := make([]byte, 0, 91)
+	s = append(s, m.OrderID...)
+	s = append(s, m.MatchID...)
+	s = append(s, uint64Bytes(m.Quantity)...)
+	s = append(s, uint64Bytes(m.Rate)...)
+	s = append(s, uint64Bytes(m.Time)...)
+	s = append(s, []byte(m.Address)...)
+	return s, nil
+}
+
+// Init is the payload for a client-originating InitRoute request.
+type Init struct {
+	signable
+	OrderID  Bytes `json:"orderid"`
+	MatchID  Bytes `json:"matchid"`
+	TxID     string `json:"txid"`
+	Vout     uint32 `json:"vout"`
+	Time     uint64 `json:"timestamp"`
+	Contract Bytes `json:"contract"`
+}
+
+var _ Signable = (*Init)(nil)
+
+// Serialize serializes the Init data.
+func (init *Init) Serialize() ([]byte, error) {
+	// Init serialization is orderid (32) + matchid (32) + txid (probably 64) +
+	// vout (4) + timestamp (8) + contract (97 ish). Sum = 205
```

Comment:
Regardless of how we decide to represent `TxID` as a field in the struct (the hash reversal isue), it could still be serialized as bytes to save 32 bytes.